File: python/611.valid-triangle-number.py


class Solution:
    def triangleNumber(self, nums: list[int]) -> int:
        nums.sort()

        # count valid triangles
        # a < b < c and a + b > c
        n = len(nums)

        count = 0

        # for a in range(n):
        #     for b in range(a + 1, n):
        #         c = bisect_left(nums, nums[a] + nums[b], b + 1, n)
        #         count += c - b - 1

        for c in range(n - 1, 1, -1):
            # optimize 1: if c side is so small that all two possble a and b sum will be a triangle
            # then we can count all possible combinations and finish,
            # there will be no greater right bound to check any more (as array is sorted)
            if nums[0] + nums[1] > nums[c]:
                # C(x, 3) = x! / (x - 3)! * 3!
                # C(x, 3) = x * (x - 1) * (x - 2) / 3!
                # x = c + 1 (index to count conversion)
                count += (c + 1) * c * (c - 1) // 6
                break

            # optimize 2: if c is so great than all two possible a and b sum will not be a triangle
            # then we can skip counting all possible combinations
            # and continue to look for smaller c side
            if nums[c - 1] + nums[c - 2] <= nums[c]:
                continue

            a, b = 0, c - 1

            while a < b:
                if nums[a] + nums[b] > nums[c]:
                    # number of indices between a and b (b - a + 1) inclusive
                    # minus one for b side of the triangle (b - a + 1 - 1)
                    count += b - a
                    b -= 1
                else:
                    a += 1

        return count

File: cpp/611.valid-triangle-number.cpp
#include <algorithm>
#include <vector>
using std::vector;

class Solution {
public:
  int triangleNumber(vector<int> &nums) {
    std::sort(nums.begin(), nums.end());
    const int n = nums.size();
    int count = 0;

    // observations (a, b, c is indexes):
    // 1) a < b < c and a + b > c: triangle
    // 2) if (c-2) + (c-1) <= c: no solutions
    // 3) if 0 + 1 > c: all possible solutions

    // for (int a = 0; a < n; a++) {
    //   for (int b = a + 1; b < n; b++) {
    //     // look for lowest c side index:
    //     // bin search for nums[a] + nums[b], starting from b + 1
    //     int c = std::lower_bound(nums.begin() + b + 1, nums.end(),
    //                              nums[a] + nums[b]) -
    //             nums.begin();
    //     count += c - b - 1;
    //   }
    // }

    for (int c = n - 1; c > 1; c--) {
      // optimize 1: 0 + 1 > c
      // if all sides less than c, will be a triangle
      // count all triplet combinations as triangle and finish
      // 5 5 5 5 9
      if (nums[0] + nums[1] > nums[c]) {
        // C(x, 3) = x! / ((x - 3)! * 3!)
        // C(x, 3) = x * (x - 1) * (x - 2) / 6
        // x = c + 1
        count += (c + 1) * c * (c - 1) / 6;
        break;
      }

      // optimize 2: (c-2) + (c-1) <= c
      // if all possible sides will not be a triangle
      // c side is very large, no possible triangle
      // look for smaller c
      // 2 2 3 5
      if (nums[c - 2] + nums[c - 1] <= nums[c]) {
        continue;
      }

      int a = 0, b = c - 1;

      while (a < b) {
        if (nums[a] + nums[b] > nums[c]) {
          // triangles: number of indices between a and b (b - a + 1)
          // minus one for b side index (b - a + 1 - 1) => (b - a)
          count += b - a;
          b--;
        } else {
          a++;
        }
      }
    }

    return count;
  }
};